Std/Data/HashMap/Basic.lean

```diff
@@ -77,37 +77,36 @@ A "load factor" of 0.75 is the usual standard for hash maps, so we return `capac
   capacity * 4 / 3
 
 /-- Constructs an empty hash map with the specified nonzero number of buckets. -/
-@[inline] def empty' (buckets := 8) (h : 0 < buckets := by decide) : Imp α β :=
-  ⟨0, .mk buckets h⟩
+@[inline] def empty' (nBuckets : Nat) (h : nBuckets.isPowerOfTwo) : Imp α β :=
+  ⟨0, .mk nBuckets h⟩
 
 /-- Constructs an empty hash map with the specified target capacity. -/
-def empty (capacity := 0) : Imp α β :=
-  let nbuckets := numBucketsForCapacity capacity
-  let n : {n : Nat // 0 < n} :=
-    if h : nbuckets = 0 then ⟨8, by decide⟩
-    else ⟨nbuckets, Nat.zero_lt_of_ne_zero h⟩
-  empty' n n.2
+def empty (capacity := 8) : Imp α β :=
+  let nBuckets := numBucketsForCapacity capacity |>.nextPowerOfTwo
+  empty' nBuckets (Nat.isPowerOfTwo_nextPowerOfTwo _)
 
-/-- Calculates the bucket index from a hash value `u`. -/
-def mkIdx {n : Nat} (h : 0 < n) (u : USize) : {u : USize // u.toNat < n} :=
-  ⟨u % n, USize.modn_lt _ h⟩
+/-- Calculates the bucket index from a `hash` value. -/
+/- Remark: we use a C implementation because this function is performance critical. -/
+@[extern c inline "(size_t)(#2) & (lean_unbox(#1) - 1)"]
```

Comment:
I'd be extremely surprised if this makes a difference in practice, and wouldn't believe it without actual benchmarks.  Leo hasn't seen any performance improvements from this change in Lean 4 either.  (It's also incorrect, because `sz` might be boxed in general.)

The enormous practical downside of this change is of course that you can no longer use std4 without compiling it.  We didn't require that before.

(If you're concerned about performance, I'd be more concerned about using hashes directly.  The Hashable instances are not particularly clever and often return results divisible by powers of two.  See also https://github.com/leanprover/lean4/pull/1840)

Reply:
> It's also incorrect, because `sz` might be boxed in general.

I'm confused by this point; there is a `lean_unbox` call, is that insufficient?

> The enormous practical downside of this change is of course that you can no longer use std4 without compiling it.

It's true! It is surprising to me that the interpreter does not fallback on the Lean implementation when it encounters an `extern` symbol with no code loaded. Maybe it should do this instead of crashing? Anyhow I will revert the change.

Reply:
> I'm confused by this point; there is a `lean_unbox` call, is that insufficient?

I'm sorry, picked the wrong terminogy here.  The `sz` argument might not be a scalar number (directly stored in the argument word using tagging), but it could be a heap-allocated bignum (where the argument stores the pointer).  In that case `lean_unbox` would return the pointer (right-shifted by one).

Reply:
That's true, we are assuming the map has fewer than `2^63` (or whatever the largest directly storable value is) entries, so perhaps `sz : USize` would be better.